python/pyADflow.py

```diff
@@ -1682,6 +1692,146 @@ def Func(Xn, CLstar):
+            else:
+                targetVals.append(0.0)
+            
+            if 'initVal' in funcDict[key]:
+                initVals.append(funcDict[key]['initVal'])
+            else:
+                initVals.append(0.0)
+
+            if 'initStep' in funcDict[key]:
+                initStep.append(funcDict[key]['initStep'])
+            else:
+                initStep.append(1e-1)
+
+        nVars = len(funcKeys)
+
+        def Func(Xn, targetVals):
+            x = self.DVGeo.getValues()
+            for i in range(nVars):
+                if dvKeys[i] in x:
+                    dvIdx = dvIndices[i]
+                    if dvIdx is not None:
+                        x[dvKeys[i]][dvIdx] = Xn[i]
+                    else:
+                        x[dvKeys[i]] = Xn[i]
+                else:
+                    x[dvKeys[i]] = Xn[i]
+
+            self.curAP.setDesignVars(x)
+            self.DVGeo.setDesignVars(x)
+
+            self.__call__(self.curAP, writeSolution=False)
+            self.curAP.adflowData.callCounter -= 1
+            funcs = {}
+            self.evalFunctions(self.curAP, funcs, evalFuncs=funcKeys)
+            F = numpy.zeros([nVars])
+            for i in range(nVars):
+                F[i] = funcs['%s_%s'%(self.curAP.name,funcKeys[i])]-targetVals[i]
+
+            return F
+
+        # Generate initial point
+        Xn = numpy.array(initVals)
+        Fn = Func(Xn, targetVals)
+
+        # Next we generate the initial jacobian if we haven't been
+        # provided with one.
+        if Jac0 is not None:
+            J = Jac0.copy()
+        else:
+            J = numpy.zeros((nVars,nVars))
+
+            # Perturb each var in succession
+            for i in range(nVars):
+                Xn[i]+=initStep[i]
+                Fpdelta = Func(Xn, targetVals)
+                J[:, i] = (Fpdelta - Fn)/initStep[i]
+                Xn[i] -= initStep[i]
+
+        # Main iteration loop
+        for jj in range(nIter):
+            if self.comm.rank == 0:
+                print ('Fn:', Fn)
+
+            # We now have a point and a jacobian...Newton's Method!
+            Xnp1 = Xn - numpy.linalg.solve(J, Fn)
+            if self.comm.rank == 0:
+                print ("Xnp1:", Xnp1)
+
+            # Solve the new Xnp1
+            Fnp1 = Func(Xnp1, targetVals)
+            if self.comm.rank == 0:
+                print ("Fnp1:", Fnp1)
+
+            # Update the jacobian using Broyden's method
+            dx = Xnp1 - Xn
+            dF = Fnp1 - Fn
+            J = J + numpy.outer((dF - numpy.dot(J, dx))/(numpy.linalg.norm(dx)**2), dx)
+
+            if self.comm.rank == 0:
+                print ("New J:", J)
+
+            # Shuffle the Fn and Xn backwards
+            Fn = Fnp1.copy()
+            Xn = Xnp1.copy()
+
+            # Check for convergence
+            if numpy.linalg.norm(Fn) < tol:
+                if self.comm.rank == 0:
+                    print ('Converged!', jj, Fn, Xn)
```

Comment:
maybe also print what jj, Fn, and Xn are? Also, Fn is just the error for each function, maybe the actual value would also be useful?

Reply:
We never have access to  the actual functions in this spot, so not super easy to do